### ISSUE TITLE
[Proposal] Support training JumpReLU SAEs

### ISSUE THREAD

**Issue by chanind** [jbloomAus/SAELens#330]
### Proposal 

We support running pre-trained JumpReLU SAEs, but we don't currently support them in training (afaik). These are a SOTA architecture so would be great if we could support this in training.

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/jbloomAus/SAELens/issues) in the repo (**required**)

**Comment by anthonyduong9** [comment]
I can work on this this week.

**Comment by chanind** [comment]
Awesome! Some more info about this

paper is here: https://arxiv.org/pdf/2407.14435

The main challenge with jumprelus is you need to implement a custom backwards pass. There's Jax code in the appendix of the paper to do this. Below is this code converted to pytorch (curtesy of Claude)

```python
import torch
import torch.nn as nn
import torch.autograd as autograd

def rectangle(x):
    return ((x > -0.5) & (x < 0.5)).to(x.dtype)

class TrainingJumpReLU(autograd.Function):
    @staticmethod
    def forward(ctx, x, threshold, bandwidth):
        ctx.save_for_backward(x, threshold, bandwidth)
        return x * (x > threshold)

    @staticmethod
    def backward(ctx, output_grad):
        x, threshold, bandwidth = ctx.saved_tensors
        x_grad = (x > threshold) * output_grad
        
        threshold_grad = (
            -(threshold / bandwidth)
            * rectangle((x - threshold) / bandwidth)
            * output_grad
        )
        
        # We return None for bandwidth_grad as it's not used in backward computation
        return x_grad, threshold_grad.sum(), None

jumprelu = TrainingJumpReLU.apply

# Usage example:
# x = torch.randn(10, requires_grad=True)
# threshold = torch.tensor(0.5, requires_grad=True)
# bandwidth = torch.tensor(0.001)
# output = jumprelu(x, threshold, bandwidth)
# output.sum().backward()
# print(x.grad)
# print(threshold.grad)
```

Other thoughts
- The paper says to keep track of `log_threshold` rather then `threshold`, so the threshold can never go negative in training. Our current jumprelu SAE implementation tracks `threshold` instead of `log_threshold`, so we'll need to resolve that somehow. Either change our jumprelu SAE implementation to use `log_threshold` instead of `threshold`, and do `log_threshold.exp()` to get the threshold again, or at the end of training a jumprelu replace the `log_threshold` with a `threshold` before saving. IMO either of these are fine, but might be worth asking @jbloomAus and/or @ArthurConmy if they have opinions.
- The paper mentions using a bandwidth of `0.001`, that's probably a good default setting
- We should make sure to add good unit tests on the `TrainingJumpReLU` to ensure the gradients it gives are correct.
- Might be worth training a basic JumpReLU SAE on gpt2 small or something similar to make sure it seems to work correctly.

**Comment by chanind** [comment]
Forgot to also add the custom heaviside step function! This also requires a custom backwards pass. Below is that code translated to torch by claude:

```python
import torch
import torch.nn as nn
import torch.autograd as autograd

def rectangle(x):
    return ((x > -0.5) & (x < 0.5)).to(x.dtype)

class DifferentiableStep(autograd.Function):
    @staticmethod
    def forward(ctx, x, threshold, bandwidth):
        ctx.save_for_backward(x, threshold, bandwidth)
        return (x > threshold).to(x.dtype)

    @staticmethod
    def backward(ctx, output_grad):
        x, threshold, bandwidth = ctx.saved_tensors
        x_grad = torch.zeros_like(output_grad)  # We don't apply STE to x input
        threshold_grad = (
            -(1.0 / bandwidth) * rectangle((x - threshold) / bandwidth) * output_grad
        )
        # We return None for bandwidth_grad as it's not used in backward computation
        return x_grad, threshold_grad.sum(), None

step = DifferentiableStep.apply

# Usage example:
# x = torch.randn(10, requires_grad=True)
# threshold = torch.tensor(0.5, requires_grad=True)
# bandwidth = torch.tensor(0.1)
# output = step(x, threshold, bandwidth)
# output.sum().backward()
# print(x.grad)
# print(threshold.grad)
```

This is used for the L0 penalty term in jumprelu training, like `penalty = lambda * step(encode(x))`

**Comment by ArthurConmy** [comment]
Here is PyTorch trainng code. On priors I'd guess Claude fucked at least one thing up: https://colab.sandbox.google.com/drive/1PlFzI_PWGTN9yCQLuBcSuPJUjgHL7GiD
